### PR TITLE
Simplify updateHead handling by using socket.unshift()

### PR DESCRIPTION
**Pull request by Nibbler999**
Simplify updateHead handling by using socket.unshift()

### CODE REVIEW

**Comment by lpinca** [comment]
This is very elegant but I think that it's better to ignore the data that is passed along with the upgrade request.
The RFC says that data transfer should start only when the handshake is complete.

**Comment by lpinca** [comment]
Anyway if maintainers want to keep the current behavior this is a nice approach.

**Comment by Nibbler999** [comment]
The same code is used to handle the upgrade response, so it's worth changing regardless.

**Comment by lpinca** [comment]
> The same code is used to handle the upgrade response, so it's worth changing regardless.

Yup, without the `unshift` 😉.

**Comment by 3rd-Eden** [comment]
I'm neutral on this change.

File: lib/WebSocket.js [hunk 1]
```diff
@@ -808,35 +807,16 @@ function establishConnection(ReceiverClass, SenderClass, socket, upgradeHead) {
   ultron.on('error', cleanupWebsocketResources.bind(this));
 
   // ensure that the upgradeHead is added to the receiver
-  function firstHandler(data) {
-    if (called || self.readyState === WebSocket.CLOSED) return;
```
Comment (Nibbler999) [inline lib/WebSocket.js]: That situation cannot arise now as we do everything in the same tick.

Comment (lpinca) [inline lib/WebSocket.js]: The `data` event is not emitted on the same tick but I guess the `data` listener is removed when `WebSocket` is closed so it should be fine.

**Pull request merged**